XIPs/xip-31-message-backup-request-system.md

```diff
@@ -0,0 +1,454 @@
+        // Perform the following steps:
+        // 1. If approve is false, delete the MessageHistoryBackupRequest from the local database and return
+        // 2. Dump all messages from the database into a correctly encoded MessageHistoryBackupFile format
+        // 3. Generate a random encryption key, nonce, and salt
+        // 4. Encrypt the file using the encryption parameters from step 2.
+        // 5. Upload the file to the provided `backupStorageProviderUploadUrl`
+        // 6. Send a message over the XMTP network containing the encryption parameters and the `backupUrl` returned from the backup storage provider
+        // 7. Delete the backup request from the database and return
+        ...
+    }
+
+    /**
+    Methods for Backup Account Files
+    **/
+    pub fn createBackupAccountFile(&self, path: String) -> Result<(), BuckupCreateError> {
+        // Create a file with a new account's private keys and return the contents of the file
+        ...
+    }
+
+    pub fn importFromBackupAccountFile(&self, path: String) -> Result<(), BackupApplyError> {
+        // Load the file from the path, spin up a temporary client with the account keys, download all messages, and create a new Message Backup File
+        // Technically we could simplify this to just load the messages directly into the database without the intermediate file, but I like the idea of reducing surface area and re-using as much code as possible from the remote option
+        ...
+    }
+}
+
+pub struct MessageHistoryBackupRequest {
+    pub requestId: String,
+    pub verificationPin: i16, // A four digit PIN that can be displayed in both the Backup Requester app and the Backup Provider app to ensure the user is responding to the correct backup request
+    pub backupStorageProviderUploadUrl: String,
+    pub status: BackupRequestStatus
+    ...
+}
+
+pub struct MessageHistoryBackupResponse {
+    pub requestId: String,
+    pub verificationPin: i16,
+    backupUrl: String,
+    encryptionKey: Vec<u8>,
+    nonce: Vec<u8>,
+    salt: Vec<u8>,
+    backupFileHash: Vec<u8>,
+    expirationTimeNs: u8
+}
+
+pub enum BackupRequestStatus {
+    Pending,
+    Expired,
+    Applied,
+    Failed
+}
+
+pub enum BackupApplyError {
+    ResponseNotFound,
+    DecryptionFailed,
+    ValidationFailed,
+}
+
+pub enum BackupCreateError {
+    RequestNotFound,
+    InvalidRequest,
+    NetworkError,
+    EncryptionError
+}
+```
+
+## Rationale
+
+There are many other potential solutions to message portability listed below. I am proposing this particular solution because it solves the key user problems stated above with the minimal amount of developer friction. Many of the alternatives require substantial cooperation from application developers, who may care more about shipping new features than supporting competing applications. Developers have strong incentives to receive backups, but limited incentives to provide them.
```

Comment:
Any potential network fees would be the responsibility of the Remote Message Backup provider as the sender?